hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java

```diff
@@ -46,36 +48,130 @@
+                this.lastAccessTime = lastAccessTime;
+                this.nextSequence = nextSequence;
+            }
+        }
+
+        static final long MC_EVENTS_WINDOW_MILLIS = TimeUnit.SECONDS.toMillis(30);
+        private final LongSupplier clock;
+        private volatile long lastMCEventsPollMillis;
+        private volatile long lastCleanupTimestamp;
+        private final ConcurrentMap<Address, LastPollRecord> lastPollRecordPerMCAddress = new ConcurrentHashMap<>();
+        private final ConcurrentArrayRingbuffer<Event> mcEvents;
+
+        MCEventStore(LongSupplier clock, ConcurrentArrayRingbuffer<Event> mcEvents) {
+            this.clock = clock;
+            this.lastMCEventsPollMillis = clock.getAsLong();
+            this.lastCleanupTimestamp = this.lastMCEventsPollMillis;
+            this.mcEvents = mcEvents;
+        }
+
+        void log(Event event) {
+            if (clock.getAsLong() - lastMCEventsPollMillis > MC_EVENTS_WINDOW_MILLIS) {
+                // ignore event and clear the queue if the last poll happened a while ago
+                onMCEventWindowExceeded();
+            } else {
+                mcEvents.add(event);
+                cleanUpLastAccessRecords();
+            }
+        }
+
+        private static boolean isOutOfTimeWindow(long nowInMillis, long subjectTimestampInMillis) {
+            return nowInMillis - subjectTimestampInMillis > MC_EVENTS_WINDOW_MILLIS;
+        }
+
+        /**
+         * Removing {@link #lastPollRecordPerMCAddress} entries older than {@link #MC_EVENTS_WINDOW_MILLIS} according to their
+         * {@code lastAccessTime}.
+         * <p>
+         * No-op if less than {@link #MC_EVENTS_WINDOW_MILLIS} millis passed since the last cleanup. This avoids too frequent
+         * unnecessary iterations on the {@link #lastPollRecordPerMCAddress} entries.
+         */
+        private void cleanUpLastAccessRecords() {
+            long now = clock.getAsLong();
+            if (!isOutOfTimeWindow(now, lastCleanupTimestamp)) {
+                return;
+            }
+            lastPollRecordPerMCAddress.entrySet().removeIf(entry -> isOutOfTimeWindow(now, entry.getValue().lastAccessTime));
+            lastCleanupTimestamp = now;
+        }
+
+        void onMCEventWindowExceeded() {
+            mcEvents.clear();
+            lastPollRecordPerMCAddress.clear();
+        }
+
+        /**
+         * @param mcRemoteAddr the address of the calling MC instance.
+         * @return the events which were added to the queue since this MC last polled, or all known events if the MC polls for the
+         * first time.
+         */
+        public List<Event> pollMCEvents(Address mcRemoteAddr) {
+            lastMCEventsPollMillis = clock.getAsLong();
+            LastPollRecord lastPollRecord = lastPollRecordPerMCAddress.get(mcRemoteAddr);
+            long sequence;
+            if (lastPollRecord == null) {
+                sequence = 0;
```

Comment:
There is one edge case related with storing sequences in `lastPollRecordPerMCAddress`. If an MC instance fails to poll events from more than 30 seconds, its sequence gets removed from the map. Then the next time it polls for the events it may receive duplicate events. This edge doesn't seem to be very critical although. WDYT?

Reply:
Well, I think after 30sec we can consider the MC as gone, but we can increase this limit to 120secs, if you agree.

Reply:
> we can increase this limit to 120secs, if you agree.

This won't give us 100% guarantee, but extending the time window should decrease the probability of the scenario. So, I'm +1 for doing this.

BTW there is also an edge case of MC restart where MC gets a new client UUID. But, luckily, MC doesn't persist WAN events (at least, now), so such situation won't do any harm.